dotcom-rendering/src/web/lib/addFontsToIframe.ts

```diff
@@ -0,0 +1,71 @@
+// The ts.dom interface for FontFaceSet does not contain the .add method
+type FontFaceSetWithAdd = FontFaceSet & {
+	add?: { (font: FontFace): void };
+};
+
+/**
+ * clone the whole style element containing the `font-face` declarations
+ * from the parent and append it to the iframe's document head.
+ */
+const cloneFontFaceElement = (
+	parentDocument: Document,
+	iframe: HTMLIFrameElement,
+) => {
+	const webfontStyleElementOnParent =
+		parentDocument.querySelector('style.webfont');
+	if (!webfontStyleElementOnParent) {
+		return;
+	}
+	const clonedStyleElement = webfontStyleElementOnParent.cloneNode(true);
+	iframe.contentWindow?.document.head.appendChild(clonedStyleElement);
+};
+
+/**
+ * Copy a subset of fonts from the parent document to an iframe.
+ * If the browser does not support the FontFace.add method,
+ * clones the style element with the "webfont" class and appends
+ * it to the iframe's document head.
+ */
+export const addFontsToIframe = (
+	parentDocument: Document,
+	iframe: HTMLIFrameElement,
+	requiredFontNames: string[],
+): void => {
+	// FontFace.add is not supported (IE), clone the element instead
+	const iframeFontFaceSet = iframe.contentDocument?.fonts as
+		| undefined
+		| FontFaceSetWithAdd;
+	if (!iframeFontFaceSet?.add) {
+		cloneFontFaceElement(parentDocument, iframe);
+		return;
+	}
+
+	// get all the fontFaces on the parent matching the list of font names
+	const requiredFonts: FontFace[] = [];
+	parentDocument.fonts.forEach((fontFace) => {
+		if (requiredFontNames.includes(fontFace.family)) {
+			requiredFonts.push(fontFace);
+		}
+	});
+
+	// add the fonts to the iframe
+	let usingAddFailed = false;
+	requiredFonts.forEach((font) => {
+		try {
+			if (usingAddFailed) {
+				return;
+			}
+			iframeFontFaceSet.add(font);
+		} catch (error) {
+			// Safari throws an InvalidModificationError on the add(font) method
+			// https://developer.mozilla.org/en-US/docs/Web/API/FontFaceSet/add#exceptions
+
+			// This may be a Safari bug - the error indicates that the FontFace is already
+			// defining in the document with a css @font-face declaration. The parent window
+			// has @font-face declarations, but the iframe does not.
+
+			usingAddFailed = true;
+			cloneFontFaceElement(parentDocument, iframe);
```

Comment:
After the first exception is thrown we set `usingAddFailed` to `true`, would this not cause any future iterations of this loop to early return on https://github.com/guardian/dotcom-rendering/pull/7533/files#diff-b9cd1532caa176968f4ffc39f3256e54078df602cdb56962580dc82a148049e9R56
thus the font does not get added to the iframe by using `.add` or by cloning (as no exception would be thrown) ?

TLDR no more fonts get added/cloned after one throws an exception?

Reply:
It will cause future iterations to early return, but this is intended. `cloneFontFaceElement` copies the whole of the `<style>` element from the parentDocument into the iframe - so if adding one font throws an exception, the whole set of fonts gets copied instead (which we only want to do once).

Or, to put it another way - if one call to `iframeFontFaceSet.add` fails, we assume rest will fail too, so just copy the whole `<style>` element instead.

Looking back at this code, I think it could benefit from some more comments and/or refactoring. It's not very clear.

Reply:
Aha! Gotcha, that makes sense now, thanks!